Hair_reredux

Vimas21 opened the pull request:
:cl: SammasC123
rscadd: Adds capilliumate and hair dye as reagents. Capilliumate grows hair, and hair dye changes a persons hair color (facial and head). Works on contact.
/:cl:

SammasC123 committed:
Hair_reredux

Literally copy pasted into a new commit, please dear god let this work
Metabolism Increase

Now it should only do it once

Albens commented:
:cl: (your ckey)
rscadd: Adds capilliumate and hair dye as reagents. Capilliumate grows hair, and hair dye changes a persons hair color (facial and head). Works on contact.
/:cl:

PsiOmegaDelta commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,75 @@
+	color = "#696969"
+
+/datum/reagent/colored_hair_dye/affect_touch(var/mob/living/carbon/human/M, var/alien, var/removed)
+	if(!alien)
```
Memory serves, species have a set of flags that declare if they have colorable hair

SierraKomodo replied:
`M.species.appearance_flags & HAS_HAIR_COLOR`

This _should_ work.

You should probably also add an `istype(M)` check as well if you don't intend this to run on all instances of `/mob/living/carbon` instead of `/mob/living/carbon/human` - The original parent proc's parameters expect carbon instead of human.

Vimas21 replied:
Do you recommend setting it to just human or keep it just as /mob/living/carbon and then add HAS_HAIR_COLOR? I think I'm getting two seperate suggestions

PsiOmegaDelta commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Other.dm:
```diff
@@ -517,3 +517,75 @@
+/datum/reagent/capilliumate/affect_touch(var/mob/living/carbon/human/M, var/alien, var/removed)
+	if(!alien)
+		M.change_hair("Very Long Hair")
+		M.change_facial_hair("Very Long Beard")
```
I'll have to look into how this is usually done, or can be done better, because strings-use like this is concerning

SierraKomodo replied:
Looking at existing code - Most instances are from vars that are set by user interaction pulled off a list of existing hairs, with one exception that has a hard coded "Bald" string.

Vimas21 replied:
So it seems like the precedence is set by having the user select it. If it's pulled off of a list, that's also just grabbing from a list of strings.

The hair datatype only contains name, icon state, and flags. I could do /datum/sprite_accessory/hair/longest.name, but I don't see how that's any different.

PsiOmegaDelta replied:
The difference is compile-time safety.
As is, if these hairstyles are renamed or removed this part will just silently break. By relying on a type/instance you'd always get the correct name or the compiler will protest if the type is removed.

SierraKomodo reviewed (changes requested):
Forgot to click the review button before leaving the comment - See above comments.

SammasC123 committed:
Fixin Some ch-ch-changes

Giving me a runtime and I'm not sure why

SammasC123 committed:
Runtime Error Fixorama

Neato

SammasC123 committed:
Hair_reredux

Literally copy pasted into a new commit, please dear god let this work
Metabolism Increase

Now it should only do it once
Fixin Some ch-ch-changes

Giving me a runtime and I'm not sure why
Runtime Error Fixorama

Neato

SammasC123 committed:
Merge remote-tracking branch 'Vimas21/hairchems_reredux' into hairchems_reredux
